pwndbg/commands/cymbol.py

```diff
@@ -0,0 +1,269 @@
+            gcc_flags = [gcc_compiler_path, "-m64"]
+        elif pwndbg.gdblib.arch.ptrsize == 4:
+            gcc_flags = [gcc_compiler_path, "-m32"]
+    else:
+        # Should we check for exceptions here?
+        gcc_flags = pwndbg.lib.gcc.which(pwndbg.gdblib.arch)
+
+    gcc_cmd = gcc_flags + gcc_extra_flags
+
+    try:
+        subprocess.run(gcc_cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, check=True)
+    except subprocess.CalledProcessError as exepction:
+        print(message.error(exepction))
+        print(message.error("Parsing failed, try to fix any syntax errors first."))
+        return None
+    except Exception as exepction:
+        print(message.error(exepction))
+        print(message.error("An error occured while generating the debug symbols."))
+        return None
+
+    return pwndbg_debug_symbols_output_file
+
+
+@PromptForOverwrite
+def add_custom_structure(custom_structure_name):
+    print(
+        message.notice("Enter your custom structure in a C header style, press Ctrl+D to save:\n")
+    )
+
+    custom_structures_source = sys.stdin.read().strip()
+    if custom_structures_source == "":
+        print(message.notice("An empty structure is entered, skipping ..."))
+        return
+
+    pwndbg_cachedir = pwndbg.lib.tempfile.cachedir("custom-symbols")
+    pwndbg_custom_structure_path = os.path.join(pwndbg_cachedir, custom_structure_name) + ".c"
+
+    with open(pwndbg_custom_structure_path, "w") as f:
+        f.write(custom_structures_source)
+
+    pwndbg_debug_symbols_output_file = generate_debug_symbols(pwndbg_custom_structure_path)
+    if not pwndbg_debug_symbols_output_file:
+        return
+
+    gdb.execute(f"add-symbol-file {pwndbg_debug_symbols_output_file}", to_string=True)
+    print(message.success("Symbols are added!"))
+
+
+@OnlyWhenStructureExists
+def edit_custom_structure(custom_structure_name, custom_structure_path):
+
+    # Lookup an editor to use for editing the custom structure.
+    editor_preference = os.getenv("EDITOR")
+    if not editor_preference:
+        editor_preference = os.getenv("VISUAL")
+    if not editor_preference:
+        editor_preference = "vi"
+
+    if cymbol_editor != "":
+        editor_preference = cymbol_editor
+
+    try:
+        subprocess.run(
+            [editor_preference, custom_structure_path],
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            check=True,
+        )
+    except Exception as exepction:
+        print(message.error("An error occured during opening the source file."))
+        print(message.error(f"Path to the custom structure: {custom_structure_path}"))
+        print(message.error("Please try to manually edit the structure."))
+        print(
+            message.error(
+                '\nTry to set a path to an editor with:\n\tset "cymbol-editor" /usr/bin/nano'
+            )
+        )
+        return
+
+    input(message.notice("Press enter to continue."))
```

Comment:
Is this needed?

Reply:
Yeah it is needed when the editor is a GUI app like vs code etc that might not be closed when the user is done with editing it. But also it is convenient for the user to have it open in the background.

Reply:
Interesting, so e.g. vim will block but vscode will not, right? Maybe we should inform user to press enter after they saved the struct, but maybe it is fine as it is

Reply:
Yeah at least when I put `export EDITOR=code`, vs code doesn't seem to block pwndbg, might need some extra flags, I will check it. But yeah it would be better to display a more informative message to the user rather than a generic 'Press enter to continue.'.